File: internal/storage/sqlite/delivery.go
package sqlite

import (
	"context"
	"database/sql"
	"errors"
	"fmt"
	"strings"
	"time"
)

type SentItem struct {
	Title  string    `json:"title"`
	URL    string    `json:"url"`
	SentAt time.Time `json:"sent_at"`
}

type Delivery struct {
	RunID       string
	Message     string
	DeliveredAt time.Time
}

func (s *Store) RecentDeliveries(ctx context.Context, limit int) ([]Delivery, error) {
	if limit <= 0 {
		limit = 2
	}
	rows, err := s.db.QueryContext(ctx, `
SELECT run_id, message, delivered_at FROM delivery
ORDER BY delivered_at DESC, id DESC
LIMIT ?`, limit)
	if err != nil {
		return nil, err
	}
	defer func() {
		_ = rows.Close()
	}()
	var deliveries []Delivery
	for rows.Next() {
		var delivery Delivery
		var deliveredAt string
		if err := rows.Scan(&delivery.RunID, &delivery.Message, &deliveredAt); err != nil {
			return nil, err
		}
		delivery.DeliveredAt, _ = time.Parse(time.RFC3339Nano, deliveredAt)
		deliveries = append(deliveries, delivery)
	}
	return deliveries, rows.Err()
}

func (s *Store) RecentSentItems(ctx context.Context, since time.Time) ([]SentItem, error) {
	rows, err := s.db.QueryContext(ctx, `
SELECT title, url, sent_at FROM sent_item
WHERE sent_at >= ?
ORDER BY sent_at DESC, id DESC`, since.UTC().Format(time.RFC3339Nano))
	if err != nil {
		return nil, err
	}
	defer func() {
		_ = rows.Close()
	}()
	var items []SentItem
	for rows.Next() {
		var item SentItem
		var sentAt string
		if err := rows.Scan(&item.Title, &item.URL, &sentAt); err != nil {
			return nil, err
		}
		item.SentAt, _ = time.Parse(time.RFC3339Nano, sentAt)
		items = append(items, item)
	}
	return items, rows.Err()
}

func (s *Store) InsertDelivery(ctx context.Context, runID string, message string, items []SentItem) ([]SentItem, error) {
	if strings.TrimSpace(runID) == "" {
		return nil, errors.New("run_id is required")
	}
	tx, err := s.db.BeginTx(ctx, nil)
	if err != nil {
		return nil, err
	}
	claim, err := tx.ExecContext(ctx, `
INSERT INTO delivery_once (run_id, message)
VALUES (?, ?)
ON CONFLICT(run_id) DO NOTHING`, runID, message)
	if err != nil {
		_ = tx.Rollback()
		return nil, err
	}
	claimed, err := claim.RowsAffected()
	if err != nil {
		_ = tx.Rollback()
		return nil, err
	}
	if claimed == 0 {
		var storedMessage string
		var deliveryID sql.NullInt64
		if err := tx.QueryRowContext(ctx, `SELECT message, delivery_id FROM delivery_once WHERE run_id = ?`, runID).Scan(&storedMessage, &deliveryID); err != nil {
			_ = tx.Rollback()
			return nil, err
		}
		if storedMessage != message {
			_ = tx.Rollback()
			return nil, errors.New("run_id was already delivered with a different message")
		}
		if !deliveryID.Valid {
			_ = tx.Rollback()
			return nil, errors.New("delivery idempotency record is incomplete")
		}
		stored, err := sentItemsForDelivery(ctx, tx, deliveryID.Int64)
		if err != nil {
			_ = tx.Rollback()
			return nil, err
		}
		if err := tx.Commit(); err != nil {
			return nil, err
		}
		return stored, nil
	}
	deliveredAt := s.now().UTC()
	result, err := tx.ExecContext(ctx, `
INSERT INTO delivery (run_id, message, delivered_at)
VALUES (?, ?, ?)`, runID, message, deliveredAt.Format(time.RFC3339Nano))
	if err != nil {
		_ = tx.Rollback()
		return nil, err
	}
	deliveryID, err := result.LastInsertId()
	if err != nil {
		_ = tx.Rollback()
		return nil, err
	}
	for i := range items {
		items[i].SentAt = deliveredAt
		if _, err := tx.ExecContext(ctx, `
INSERT INTO sent_item (delivery_id, run_id, title, url, title_key, sent_at)
VALUES (?, ?, ?, ?, ?, ?)`,
			deliveryID, runID, items[i].Title, items[i].URL, NormalizeTitleKey(items[i].Title), deliveredAt.Format(time.RFC3339Nano)); err != nil {
			_ = tx.Rollback()
			return nil, err
		}
	}
	if _, err := tx.ExecContext(ctx, `UPDATE delivery_once SET delivery_id = ? WHERE run_id = ?`, deliveryID, runID); err != nil {
		_ = tx.Rollback()
		return nil, err
	}
	if err := tx.Commit(); err != nil {
		return nil, err
	}
	return items, nil
}

func sentItemsForDelivery(ctx context.Context, tx *sql.Tx, deliveryID int64) ([]SentItem, error) {
	rows, err := tx.QueryContext(ctx, `SELECT title, url, sent_at FROM sent_item WHERE delivery_id = ? ORDER BY id`, deliveryID)
	if err != nil {
		return nil, err
	}
	defer func() { _ = rows.Close() }()
	var items []SentItem
	for rows.Next() {
		var item SentItem
		var sentAt string
		if err := rows.Scan(&item.Title, &item.URL, &sentAt); err != nil {
			return nil, err
		}
		item.SentAt, _ = time.Parse(time.RFC3339Nano, sentAt)
		items = append(items, item)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("read idempotent delivery items: %w", err)
	}
	return items, nil
}

func NormalizeTitleKey(text string) string {
	text = strings.ToLower(strings.TrimSpace(text))
	var b strings.Builder
	previousSpace := false
	for _, r := range text {
		switch {
		case r >= 'a' && r <= 'z', r >= '0' && r <= '9':
			b.WriteRune(r)
			previousSpace = false
		default:
			if !previousSpace {
				b.WriteByte(' ')
				previousSpace = true
			}
		}
	}
	return strings.Join(strings.Fields(b.String()), " ")
}

File: crates/siftwire/src/storage/delivery.rs
use anyhow::{Context, Result, anyhow, ensure};
use chrono::{DateTime, Utc};
use rusqlite::{OptionalExtension, Transaction, params};

use super::{
    Delivery, DeliveryInsertError, Store, StoredSentItem, format_timestamp, parse_timestamp,
};

impl Store {
    /// Returns recent deliveries newest first.
    ///
    /// The limit must be positive.
    ///
    /// # Errors
    ///
    /// Returns an error when the limit is non-positive or a stored row is invalid.
    pub fn recent_deliveries(&self, limit: i64) -> Result<Vec<Delivery>> {
        ensure!(limit > 0, "delivery limit must be positive");
        let mut statement = self
            .connection
            .prepare(
                "SELECT run_id, message, delivered_at FROM delivery \
                 ORDER BY delivered_at DESC, id DESC LIMIT ?1",
            )
            .context("prepare recent delivery query")?;
        let rows = statement
            .query_map([limit], |row| {
                Ok((
                    row.get::<_, String>(0)?,
                    row.get::<_, String>(1)?,
                    row.get::<_, String>(2)?,
                ))
            })
            .context("query recent deliveries")?;
        rows.map(|row| {
            let (run_id, message, delivered_at) = row.context("read delivery row")?;
            Ok(Delivery {
                run_id,
                message,
                delivered_at: parse_timestamp(&delivered_at)?,
            })
        })
        .collect()
    }

    /// Returns sent items at or after the supplied UTC timestamp, newest first.
    ///
    /// # Errors
    ///
    /// Returns an error when `SQLite` cannot read a sent item row.
    pub fn recent_sent_items(&self, since: DateTime<Utc>) -> Result<Vec<StoredSentItem>> {
        let mut statement = self
            .connection
            .prepare(
                "SELECT title, url, kind, sent_at FROM sent_item \
                 WHERE sent_at >= ?1 AND kind != 'sports_schedule' \
                 ORDER BY sent_at DESC, id DESC",
            )
            .context("prepare recent sent item query")?;
        let rows = statement
            .query_map([format_timestamp(since)], |row| {
                Ok((
                    row.get::<_, String>(0)?,
                    row.get::<_, String>(1)?,
                    row.get::<_, String>(2)?,
                    row.get::<_, String>(3)?,
                ))
            })
            .context("query recent sent items")?;
        rows.map(|row| {
            let (title, url, kind, sent_at) = row.context("read sent item row")?;
            Ok(StoredSentItem {
                title,
                url,
                kind,
                sent_at: parse_timestamp(&sent_at)?,
            })
        })
        .collect()
    }

    /// Atomically records one idempotent delivery and its sent items.
    ///
    /// Repeating a run identifier with the same message returns the original items.
    /// Repeating it with a different message returns [`DeliveryInsertError::Conflict`].
    ///
    /// # Errors
    ///
    /// Returns a conflict for a changed message, or a storage error when validation or `SQLite`
    /// work fails.
    pub fn insert_delivery(
        &self,
        run_id: &str,
        message: &str,
        mut items: Vec<StoredSentItem>,
    ) -> Result<Vec<StoredSentItem>, DeliveryInsertError> {
        if run_id.trim().is_empty() {
            return Err(anyhow!("run_id is required").into());
        }
        let transaction = self
            .connection
            .unchecked_transaction()
            .context("begin delivery insert")?;
        let claimed = transaction
            .execute(
                "INSERT INTO delivery_once (run_id, message) VALUES (?1, ?2) \
                 ON CONFLICT(run_id) DO NOTHING",
                params![run_id, message],
            )
            .with_context(|| format!("claim delivery run {run_id}"))?;
        if claimed == 0 {
            return existing_delivery(transaction, run_id, message);
        }
        let delivered_at = self.timestamp();
        transaction
            .execute(
                "INSERT INTO delivery (run_id, message, delivered_at) VALUES (?1, ?2, ?3)",
                params![run_id, message, format_timestamp(delivered_at)],
            )
            .with_context(|| format!("insert delivery for run {run_id}"))?;
        let delivery_id = transaction.last_insert_rowid();
        for item in &mut items {
            item.sent_at = delivered_at;
            insert_sent_item(&transaction, delivery_id, run_id, item)?;
        }
        transaction
            .execute(
                "UPDATE delivery_once SET delivery_id = ?1 WHERE run_id = ?2",
                params![delivery_id, run_id],
            )
            .with_context(|| format!("complete delivery claim for run {run_id}"))?;
        transaction
            .commit()
            .with_context(|| format!("commit delivery for run {run_id}"))?;
        Ok(items)
    }
}

fn existing_delivery(
    transaction: Transaction<'_>,
    run_id: &str,
    message: &str,
) -> Result<Vec<StoredSentItem>, DeliveryInsertError> {
    let stored = transaction
        .query_row(
            "SELECT message, delivery_id FROM delivery_once WHERE run_id = ?1",
            [run_id],
            |row| Ok((row.get::<_, String>(0)?, row.get::<_, Option<i64>>(1)?)),
        )
        .optional()
        .with_context(|| format!("read delivery claim for run {run_id}"))?
        .ok_or_else(|| anyhow!("delivery idempotency record disappeared for run {run_id}"))?;
    if stored.0 != message {
        return Err(DeliveryInsertError::Conflict);
    }
    let delivery_id = stored
        .1
        .ok_or_else(|| anyhow!("delivery idempotency record is incomplete"))?;
    let items = sent_items_for_delivery(&transaction, delivery_id)?;
    transaction
        .commit()
        .with_context(|| format!("commit idempotent delivery for run {run_id}"))?;
    Ok(items)
}

fn insert_sent_item(
    transaction: &Transaction<'_>,
    delivery_id: i64,
    run_id: &str,
    item: &StoredSentItem,
) -> Result<()> {
    transaction
        .execute(
            "INSERT INTO sent_item \
             (delivery_id, run_id, title, url, kind, title_key, sent_at) \
             VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7)",
            params![
                delivery_id,
                run_id,
                item.title,
                item.url,
                item.kind,
                super::normalize_title_key(&item.title),
                format_timestamp(item.sent_at),
            ],
        )
        .with_context(|| format!("insert sent item for run {run_id}"))?;
    Ok(())
}

fn sent_items_for_delivery(
    transaction: &Transaction<'_>,
    delivery_id: i64,
) -> Result<Vec<StoredSentItem>> {
    let mut statement = transaction
        .prepare(
            "SELECT title, url, kind, sent_at FROM sent_item WHERE delivery_id = ?1 ORDER BY id",
        )
        .context("prepare idempotent delivery item query")?;
    let rows = statement
        .query_map([delivery_id], |row| {
            Ok((
                row.get::<_, String>(0)?,
                row.get::<_, String>(1)?,
                row.get::<_, String>(2)?,
                row.get::<_, String>(3)?,
            ))
        })
        .context("query idempotent delivery items")?;
    rows.map(|row| {
        let (title, url, kind, sent_at) = row.context("read idempotent delivery item")?;
        Ok(StoredSentItem {
            title,
            url,
            kind,
            sent_at: parse_timestamp(&sent_at)?,
        })
    })
    .collect()
}

/// Produces the normalized key used for sent-title deduplication.
#[must_use]
pub fn normalize_title_key(text: &str) -> String {
    let mut normalized = String::new();
    let mut previous_space = false;
    for character in text.trim().to_lowercase().chars() {
        if character.is_ascii_lowercase() || character.is_ascii_digit() {
            normalized.push(character);
            previous_space = false;
        } else if !previous_space {
            normalized.push(' ');
            previous_space = true;
        }
    }
    normalized.split_whitespace().collect::<Vec<_>>().join(" ")
}
